File: java/src/main/java/problems101_200/PascalsTriangle.java
package problems101_200;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

enum PascalsTriangle {
    DEFAULT {
        @Override
        List<List<Integer>> generate(int numRows) {
            List<List<Integer>> rows = new ArrayList<>(numRows);
            rows.add(Collections.singletonList(1));
            for (int i = 1; i < numRows; i++) {
                int numColumns = i + 1;
                List<Integer> columns = new ArrayList<>(numColumns);
                for (int j = 0; j < numColumns; j++) {
                    if (j == 0 || j == numColumns - 1) {
                        columns.add(1);
                        continue;
                    }
                    List<Integer> prevColumns = rows.get(i - 1);
                    columns.add(prevColumns.get(j - 1) + prevColumns.get(j));
                }
                rows.add(columns);
            }
            return rows;
        }
    };

    abstract List<List<Integer>> generate(int numRows);
}

File: kotlin/src/main/kotlin/problems101_200/PascalsTriangle.kt
package problems101_200

enum class PascalsTriangle {
    DEFAULT {
        override fun generate(numRows: Int): List<List<Int>> {
            val rows = ArrayList<List<Int>>(numRows)
            rows += listOf(1)
            for (i in 1 until numRows) {
                val numColumns = i + 1
                val columns = ArrayList<Int>(numColumns)
                for (j in 0 until numColumns) {
                    if (j == 0 || j == numColumns - 1) {
                        columns += 1
                        continue
                    }
                    val prevColumns = rows[i - 1]
                    columns += prevColumns[j - 1] + prevColumns[j]
                }
                rows += columns
            }
            return rows
        }
    },
    ;

    abstract fun generate(numRows: Int): List<List<Int>>
}
